src/db/gc/mod.rs

```diff
@@ -0,0 +1,309 @@
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let marked = mem::take(&mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
+    /// than chain finality.
+    /// * `interval` - GC Interval to avoid constantly consuming node's resources.
+    ///
+    /// NOTE: This currently does not take into account the fact that we might be starting the node
+    /// using CAR-backed storage with a snapshot, for implementation simplicity.
+    pub async fn gc_loop(&mut self, interval: Duration) -> anyhow::Result<()> {
+        loop {
+            self.gc_workflow(interval).await?
+        }
+    }
+
+    // This function yields to the main GC loop if the conditions are not met for execution of the
+    // next step.
+    async fn gc_workflow(&mut self, interval: Duration) -> anyhow::Result<()> {
+        let depth = self.depth;
+        let tipset = (self.get_heaviest_tipset)();
+        let current_epoch = tipset.epoch();
+        // Don't run the GC if there aren't enough state-roots yet. Sleep and yield to the main loop
+        // in order to refresh the heaviest tipset value.
+        if depth > current_epoch {
+            time::sleep(self.block_time * (depth - current_epoch) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        if self.marked.is_empty() {
+            info!("populate keys for GC");
+            self.populate()?;
+            self.epoch_marked = current_epoch;
+        }
+
+        let epochs_since_marked = current_epoch - self.epoch_marked;
+        // Don't proceed with next steps until we advance at least `depth` epochs. Sleep and yield
+        // to the main loop in order to refresh the heaviest tipset value.
+        if epochs_since_marked < depth {
+            time::sleep(self.block_time * (depth - epochs_since_marked) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        info!("filter keys for GC");
+        self.filter(tipset, depth).await?;
+
+        info!("GC sweep");
+        self.sweep()?;
+
+        // Make sure we don't run the GC too often.
+        time::sleep(interval).await;
+
+        anyhow::Ok(())
+    }
+}
+#[cfg(test)]
+mod test {
+    use crate::blocks::{BlockHeader, Tipset};
+    use crate::chain::ChainStore;
+
+    use crate::db::{GarbageCollectable, MarkAndSweep, MemoryDB};
+    use crate::message_pool::test_provider::{mock_block, mock_block_with_parents};
+    use crate::networks::ChainConfig;
+
+    use crate::utils::db::CborStoreExt;
+
+    use core::time::Duration;
+
+    use crate::shim::clock::ChainEpoch;
+    use fvm_ipld_blockstore::Blockstore;
+    use std::sync::Arc;
+
```

Comment:
How about a quickcheck test that GCs a blockchain with N blocks and a depth of D?

Reply:
What do you mean by that? As in we let quickcheck generate N and D for us?

Reply:
It would have to be within a certain range, but sure, we can do it.

Reply:
I have looked into that, the issue is that quickcheck does not currently support async functions. There is a `quckcheck_async` library, but I have failed to make it work so far.

Reply:
Done.